### PR TITLE
Handling of Secrets Data for Terraform Providers

### PR DESCRIPTION
**Pull request by lakshmimsft**
Design doc for Handling of Secrets Data for Terraform Providers

### CODE REVIEW

File: recipe/2024-04-terraform-provider-secrets.md [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+N/A
+
+### API design (if applicable)
+N/A
+
+### CLI Design (if applicable)
+N/A
+
+### Implementation Details
+####Recipes (Engine and Driver)
```
Comment (rynowak) [inline recipe/2024-04-terraform-provider-secrets.md:160]: I think something went wrong with the formatting here without a space between `####` and `Recipes`

File: recipe/2024-04-terraform-provider-secrets.md [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+type SecretReference {
+  source: string;		// SecretStore Id					
+  key: string;
+}
+
+type RecipeSecretIDs struct {
+  moduleSecret string                  // SecretStoreId associated with git private terraform repo source
+  providerSecrets map[string]map[string]SecretReference  // map[providerName or provider.aliasName][secret_name]SecretReference
+  envSecrets map[string]SecretReference // map[environmentvariablename]SecretReference 
+}
```
Comment (rynowak) [inline recipe/2024-04-terraform-provider-secrets.md:174]: Is this type specific to Terraform?

There's a tradeoff here between:

- The recipe engine can provide drivers with access to secrets.
- The recipe engine knows what the secrets will be used for.

File: recipe/2024-04-terraform-provider-secrets.md [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+### API design (if applicable)
+N/A
+
+### CLI Design (if applicable)
+N/A
+
+### Implementation Details
+####Recipes (Engine and Driver)
+
+```
```
Comment (rynowak) [inline recipe/2024-04-terraform-provider-secrets.md:162]:
```suggestion
```go
```

File: recipe/2024-04-terraform-provider-secrets.md [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+### Detailed Design
+
+The FindSecretIds function in the DriverWithSecrets interface is updated to return a struct of RecipeSecretIds, which will hold secret Ids for a Terraform module as before and in addition, also populate secret Ids for provider configuration and environment variables, iterating through the input environment recipe configuration. 
+
+The LoadSecrets method in the SecretLoader interface will also be updated to LoadRecipeSecrets and return struct RecipeSecrets which will contain secrets for module, providers and environment variables.
+
+#### Advantages (of each option considered)
+Retrieving secrets in the Engine and passing them into the Driver was discussed positively as a way forward. Looking ahead with Containerization, the Driver will have all the information it needs to be able to be spin off into a separate process.
+
+#### Disadvantages (of each option considered)
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:147]: There can be multiple secrets registered across different providers, but a particular recipe deployment might (most likely) won't require all of them. Is there a downside to passing all the extra information? Is there an alternative option?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+n/a
+
+## Design
+
+### High Level Design
+
+The design for handling secrets for Private Terraform modules established the basis for data flow between Engine and Driver components in the codebase. For Terraform Providers, we will continue to maintain the flow where secrets are retrieved in the Engine and passed into the Driver in the Execute() call. 
+The following diagram describes the updated data flow. The calls in green highlight updates to existing code.
+
+
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:134]: How does the driver differentiate between secrets needed for terraform module auth vs provider auth?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+
+// FindSecretIDs will be updated in current DriverWithSecrets interface
+type DriverWithSecrets interface {
+	// Driver is an interface to implement recipe deployment and recipe resources deletion.
+	Driver
+
+	// FindSecretIDs gets the secret store resource IDs for the recipe including module, provider, envSecrets.
+	FindSecretIDs(ctx context.Context, config recipes.Configuration, definition recipes.EnvironmentDefinition) (**RecipeSecretIDs**, error)
```
Comment (rynowak) [inline recipe/2024-04-terraform-provider-secrets.md:189]: What would we lose if we returned `([]string, error)` from this function?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+	Definition recipes.EnvironmentDefinition
+
+	// Secrets specifies the module authentication information stored in the secret store.
+	Secrets **RecipeSecrets**
+}
+```
+
+### Error Handling
+
+If there is an error when iterating through the recipe configuration or calling ListSecrets API, error will be returned. These errors will not include sensitive data on secret keys or values.
```
Comment (rynowak) [inline recipe/2024-04-terraform-provider-secrets.md:220]: Do we need any special error handling or specific error messages reported in this case?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+
+## Test plan
+
+Testing will include unit tests on all new functions created and a
+functional Test will be created to test this functionality e2e.
```
Comment (rynowak) [inline recipe/2024-04-terraform-provider-secrets.md:225]: It feels like we could have a negative test as well for the "secret not found" case.

File: recipe/2024-04-terraform-provider-secrets.md [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+
+## Security
+
+Secret data will be held in memory and be passed to the Driver which will inject these values into the Terraform configuration in working directory for Terraform. This directory is created when a Terraform recipe is set to be deployed and deleted once deployment of the Terraform recipe is completed..
+
+Secret data will not be persisted and secret data will not be logged by Radius.
+Terraform recipe authors should, in general, set the sensitive flag for any input and output variables that contain sensitive data.
+
```
Comment (rynowak) [inline recipe/2024-04-terraform-provider-secrets.md:233]: 👍

File: recipe/2024-04-terraform-provider-secrets.md [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+#### Proposed Option
+N/A
+
+### API design (if applicable)
+N/A
+
+### CLI Design (if applicable)
+N/A
+
+### Implementation Details
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:159]: Could you highlight parts that will be new after this proposed implementation?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+## Design
+
+### High Level Design
+
+The design for handling secrets for Private Terraform modules established the basis for data flow between Engine and Driver components in the codebase. For Terraform Providers, we will continue to maintain the flow where secrets are retrieved in the Engine and passed into the Driver in the Execute() call. 
+The following diagram describes the updated data flow. The calls in green highlight updates to existing code.
+
+
+![alt text](2024-04-terraform-provider-secrets.png)
+
```
Comment (rynowak) [inline recipe/2024-04-terraform-provider-secrets.md:136]: Love the diagram. BTW if it's easier in the future you can embed mermaid diagrams in markdown. Sequence diagram support is really good.

https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/

File: recipe/2024-04-terraform-provider-secrets.md [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+      terraform: {
+        authentication: {
+          ...
+        }
+        providers: {
+          azurerm: [
+            {
+              subscriptionid: 1234
+              tenant_id: '745fg88bf-86f1-41af'
++              secrets: { // Individual Secrets from SecretStore
```
Comment (jasonviviano) [inline recipe/2024-04-terraform-provider-secrets.md:70]: Are there any `rad CLI` commands such as `rad recipe show` or `rad env show` that need to display any information to users when a secret is registered? Such as letting a user know that secrets are configured or is that something users would have to find through the Bicep file definition?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+
+```
+Note: DataModel in [Design Document to support multiple Terraform Providers](https://github.com/radius-project/design-notes/pull/39/files)
+DataModel in Design doc describes SecretReference:
+type SecretReference {
+  source: string;		// SecretStore Id					
+  key: string;
+}
+
+type RecipeSecretIDs struct {
```
Comment (sk593) [inline recipe/2024-04-terraform-provider-secrets.md:170]: This and `RecipeSecrets` seem to have similar fields with slightly different types. Is it possible to combine these two or are the use cases too different to do that?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+
+
+![alt text](2024-04-terraform-provider-secrets.png)
+
+
+### Detailed Design
+
+The FindSecretIds function in the DriverWithSecrets interface is updated to return a struct of RecipeSecretIds, which will hold secret Ids for a Terraform module as before and in addition, also populate secret Ids for provider configuration and environment variables, iterating through the input environment recipe configuration. 
+
+The LoadSecrets method in the SecretLoader interface will also be updated to LoadRecipeSecrets and return struct RecipeSecrets which will contain secrets for module, providers and environment variables.
```
Comment (rynowak) [inline recipe/2024-04-terraform-provider-secrets.md:142]: Do other drivers have concepts like `module`, `providers`, `env-vars`?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+Note: DataModel in [Design Document to support multiple Terraform Providers](https://github.com/radius-project/design-notes/pull/39/files)
+DataModel in Design doc describes SecretReference:
+type SecretReference {
+  source: string;		// SecretStore Id					
+  key: string;
+}
+
+type RecipeSecretIDs struct {
+  moduleSecret string                  // SecretStoreId associated with git private terraform repo source
+  providerSecrets map[string]map[string]SecretReference  // map[providerName or provider.aliasName][secret_name]SecretReference
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:172]: Given the key to this map is provider name, what would it look like for private module secrets?

Comment (kachawla) [reply recipe/2024-04-terraform-provider-secrets.md:172]: Nvm. I see module secret has its own field above.

File: recipe/2024-04-terraform-provider-secrets.md [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+  envSecrets map[string]SecretReference // map[environmentvariablename]SecretReference 
+}
+
+type RecipeSecrets struct {
+  moduleSecrets v20231001preview.SecretStoresClientListSecretsResponse
+  providerSecrets map[string]map[string]string   // map[providerName/provider.aliasName][secret_name]value
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:178]: I think this explains my other question earlier as well.

Why is the data type different for `moduleSecret` and `providerSecrets` - `SecretStoresClientListSecretsResponse` vs `string`?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+	FindSecretIDs(ctx context.Context, config recipes.Configuration, definition recipes.EnvironmentDefinition) (**RecipeSecretIDs**, error)
+}
+
+LoadSecrets in SecretsLoader interface will be updated to LoadRecipeSecrets(), It will internally call functions which will populate resolved secrets based on the Ids in RecipeSecretIDs{}. The result is combined as RecipeSecrets{} which is passed as BaseOptions.Secrets to the Driver in Execute() call.
+
+LoadRecipeSecrets(RecipeSecretIDs, ...)(RecipeSecrets, error){
+  ...
+  loadModuleSecrets()
+  loadProviderSecrets()
+  loadEnvSecrets()
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:198]: Could you share more details on why these are needed as separate functions?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 18]
```diff
@@ -0,0 +1,259 @@
...
+	Definition recipes.EnvironmentDefinition
+
+	// Secrets specifies the module authentication information stored in the secret store.
+	Secrets **RecipeSecrets**
+}
+```
+
+### Error Handling
+
+If there is an error when iterating through the recipe configuration or calling ListSecrets API, error will be returned. These errors will not include sensitive data on secret keys or values.
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:220]: Will there be any new error types?

Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:220]: > These errors will not include sensitive data on secret keys or values

How are we ensuring this?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+
+## Test plan
+
+Testing will include unit tests on all new functions created and a
+functional Test will be created to test this functionality e2e.
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:225]: Curious to hear more on which provider are we planning to use for functional test?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 20]
```diff
@@ -0,0 +1,259 @@
...
+
+## Security
+
+Secret data will be held in memory and be passed to the Driver which will inject these values into the Terraform configuration in working directory for Terraform. This directory is created when a Terraform recipe is set to be deployed and deleted once deployment of the Terraform recipe is completed..
+
+Secret data will not be persisted and secret data will not be logged by Radius.
+Terraform recipe authors should, in general, set the sensitive flag for any input and output variables that contain sensitive data.
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:232]: > Terraform recipe authors should, in general, set the sensitive flag for any input and output variables that contain sensitive data.

Is this relevant to this work though? We aren't really working with variables here.

File: recipe/2024-04-terraform-provider-secrets.md [hunk 21]
```diff
@@ -0,0 +1,259 @@
...
+	// Driver is an interface to implement recipe deployment and recipe resources deletion.
+	Driver
+
+	// FindSecretIDs gets the secret store resource IDs for the recipe including module, provider, envSecrets.
+	FindSecretIDs(ctx context.Context, config recipes.Configuration, definition recipes.EnvironmentDefinition) (**RecipeSecretIDs**, error)
+}
+
+LoadSecrets in SecretsLoader interface will be updated to LoadRecipeSecrets(), It will internally call functions which will populate resolved secrets based on the Ids in RecipeSecretIDs{}. The result is combined as RecipeSecrets{} which is passed as BaseOptions.Secrets to the Driver in Execute() call.
+
+LoadRecipeSecrets(RecipeSecretIDs, ...)(RecipeSecrets, error){
```
Comment (rynowak) [inline recipe/2024-04-terraform-provider-secrets.md:194]: Suggestion for this.

```go 
LoadRecipeSecrets(ids []string)(map[string]map[string]string, error){}
```

OR

```go
LoadRecipeSecrets(secrets []SecretReference)(map[SecretReference]map[string]string, error){}
```

This would simplify the contract, and make the driver responsible for understanding its own schema, not the engine.

Comment (rynowak) [reply recipe/2024-04-terraform-provider-secrets.md:194]: Note: discussed in the meeting and agreed this was a good simplification.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 8 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md:225]: I will update test plan in doc to include this test.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 4 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md:145]: This will be discussed again based on input from PMs and their discussions with customers.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 10 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md:157]: I cleaned it up, it should be clearer. Pls take a look.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 12 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md:70]: Per discussion, this is something PMs may want to think about ahead.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 13 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md]: They were not to be combined. The function signatures changed after design discussion. Pls take a look.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 6 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md]: updated this currently to map[string]string. Also updated 'Alternatives Considered' section to describe we could go with []string.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 7 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md:193]: error messages will be similar to: eg.:  "failed to fetch secrets from the secret store resource id %s for Terraform recipe %s deployment: %s" ..

File: recipe/2024-04-terraform-provider-secrets.md (hunk 14 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md]: updated functions in design doc so we are not mirroring these constructs when handling secret data

File: recipe/2024-04-terraform-provider-secrets.md (hunk 16 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md]: updated functions in design doc per discussion.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 15 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md]: functions are now updated in design doc

File: recipe/2024-04-terraform-provider-secrets.md (hunk 18 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md:193]: Added note in test plan to add test for failure and verify the same.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 21 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md]: Updated document as mentioned. Also kept existing function name LoadSecrets(). Pls review.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 20 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md:207]: This was a general statement but I had a usecase in mind. i will try a test with an example and write back.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 19 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md]: One thought was we could use the PostGres provider and store some of the provider configuration information in a secretStore and test with that. Let me know if that sounds okay/other thoughts.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 4 above)
Comment (kachawla) [reply recipe/2024-04-terraform-provider-secrets.md:145]: > There can be multiple secrets registered across different providers, but a particular recipe deployment might (most likely) won't require all of them.

For my previous point above, can you explicitly call out this behavior in the doc?

> This will be discussed again based on input from PMs and their discussions with customers.

Let's write what specifically will we get from PMs and how that helps with the design, while the context of the discussion is still fresh in our mind. :)

File: recipe/2024-04-terraform-provider-secrets.md (hunk 10 above)
Comment (kachawla) [reply recipe/2024-04-terraform-provider-secrets.md:157]: Doesn't seem to be reflected on the PR, was the update pushed?

File: recipe/2024-04-terraform-provider-secrets.md (hunk 12 above)
Comment (kachawla) [reply recipe/2024-04-terraform-provider-secrets.md:70]: @lakshmimsft was there an action item that came out of this?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 22]
```diff
@@ -0,0 +1,234 @@
...
+N/A
+
+### API design (if applicable)
+N/A
+
+### CLI Design (if applicable)
+N/A
+
+### Implementation Details
+###Recipes (Engine and Driver)
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:158]: Which part of the changes below will go into engine? Can you separate out engine and driver changes?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 23]
```diff
@@ -0,0 +1,234 @@
...
+
+TBD
+
+## Alternatives considered
+We could also update FindSecretIds() to return a list of SecretStoreIds and the LoadSecrets function retrieves all secrets for the SecretStore and passes the data to the driver.
+We decided to limit the secret data sent to the Driver to only what the Driver will need and are going with the current design. This can be updated in the future if needed.
+
+
+## Design Review Notes
+1. Update design per discussion.Design can be more generic and need not follow Env ConfigStructure.
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:233]: Can you call out specific design decisions that were made?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 24]
```diff
@@ -0,0 +1,234 @@
...
+
+## Development plan
+Updates to Engine, Driver function call - (under 0.5 Sprint)
+Unit Tests + Functional Test - (under 0.5 Sprint)
+
+Total: 1 Sprint
+
+## Open Questions
+
+TBD
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:225]: Why is this TBD? :)

File: recipe/2024-04-terraform-provider-secrets.md (hunk 4 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md:145]: I had noted it under Design Review Notes, also added to Open Questions (we probably need it in only one section). I think I need more information about potential scenarios before writing more about it.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 10 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md:157]: yes, missed it, updated now. Pls check

File: recipe/2024-04-terraform-provider-secrets.md (hunk 12 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md:70]: tagging @jasonviviano for any further thoughts

File: recipe/2024-04-terraform-provider-secrets.md [hunk 25]
```diff
@@ -0,0 +1,260 @@
...
+> **Issue Reference:** <!-- (If appropriate) Reference an existing issue that describes the feature or bug. -->
+https://github.com/radius-project/radius/issues/6539
+
+### Goals
+
+Describe data flow, specific function calls when using secrets to configure Terraform Providers, building on design in the referred design documents.
+
+### Non goals
+
+The document is focussed on handling secrets for Terraform Providers, with Applications.Core/secretStores being the underlying resource for the secret store (which today supports Kubernetes secrets). Other sources and types of secretsStores are out of scope for this document.
```
Comment (rynowak) [inline recipe/2024-04-terraform-provider-secrets.md:34]:
```suggestion
The document is focussed on handling secrets for Terraform Providers, with `Applications.Core/secretStores` being the underlying resource for the secret store (which today supports Kubernetes secrets). Other sources and types of `secretsStores` are out of scope for this document. 
```

File: recipe/2024-04-terraform-provider-secrets.md [hunk 26]
```diff
@@ -0,0 +1,260 @@
...
+N/A
+
+### API design (if applicable)
+N/A
+
+### CLI Design (if applicable)
+N/A
+
+### Implementation Details
+###Recipes (Engine and Driver)
```
Comment (rynowak) [inline recipe/2024-04-terraform-provider-secrets.md:158]:
```suggestion
### Recipes (Engine and Driver)
```

File: recipe/2024-04-terraform-provider-secrets.md [hunk 27]
```diff
@@ -0,0 +1,260 @@
...
+	FindSecretIDs(ctx context.Context, config recipes.Configuration, definition recipes.EnvironmentDefinition) (string, error)
+}
+
+//*** NEW ***//
+
+// FindSecretIDs will be updated in current DriverWithSecrets interface to now return map of secretStoreIds and keys
+type DriverWithSecrets interface {
+  ...
+	// FindSecretIDs gets the secretStore resource IDs and keys for the recipe including module, provider, envSecrets.
+	FindSecretIDs(ctx context.Context, config recipes.Configuration, definition recipes.EnvironmentDefinition) (map[string]string, error)
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:178]: A suggestion for better readability here, making these output parameters named would help, for example, `resourceIdSecretKeyMap` in this case.

Comment (rynowak) [inline recipe/2024-04-terraform-provider-secrets.md:178]: What is the key of of the map and what is the value? I know it's strings, but what do they represent?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 28]
```diff
@@ -0,0 +1,260 @@
...
+type SecretsLoader interface {
+	LoadSecrets(map[string]string, ...)(map[string]map[string]string, error){ // returns secretStoreId[key]value
+}
+
+// When calling driver.Execute we pass in BaseOptions which is structured as:
+type BaseOptions struct {
+   ...
+
+	// Secrets specifies the module authentication information stored in the secret store.
+	Secrets map[string]map[string]string
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:212]: I think it came up in discussion so just want to confirm - were you able to confirm that this schema fits all use cases (module. provider, env vars)?

File: recipe/2024-04-terraform-provider-secrets.md (hunk 4 above)
Comment (sylvainsf) [reply recipe/2024-04-terraform-provider-secrets.md:145]: If we are only going to have it in one section updating the source of truth is better than a notes section which I would consider to be the right place for meta observations vs changes.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 28 above)
Comment (rynowak) [inline recipe/2024-04-terraform-provider-secrets.md:212]: It would be great to call out the keys and values in the docs (and in the review).

eg:

> Secrets is a map from ??? to ???

File: recipe/2024-04-terraform-provider-secrets.md (hunk 7 above)
Comment (rynowak) [reply recipe/2024-04-terraform-provider-secrets.md:218]: I think that's a good enough error message 👍

File: recipe/2024-04-terraform-provider-secrets.md (hunk 12 above)
Comment (sylvainsf) [reply recipe/2024-04-terraform-provider-secrets.md:70]: I'd like PMs to create an user story around a CLI experience for this as a separate work item.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 27 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md]: Corrected it: updated to map[string][]string // map[SecretStoreId]ListOfKeysInSecretStoreId.

Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md]: updated per comment. kept name secretIDs, added comment to describe structure.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 28 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md]: updated document

Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md]: the answer right now is yes. this should get us the data we need.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 24 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md]: updated

File: recipe/2024-04-terraform-provider-secrets.md (hunk 23 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md:260]: updated with some more detail. let me know if the level of detail is okay. Previous versions of the document are also available for users to see.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 18 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md:220]: no new error type has come up yet. Will update this note and document if that changes

File: recipe/2024-04-terraform-provider-secrets.md (hunk 4 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md:148]: Updated the Detail Design section of document. Pls review.

File: recipe/2024-04-terraform-provider-secrets.md [hunk 29]
```diff
@@ -0,0 +1,263 @@
+# Handling of Secrets Data for Terraform Providers
+
+* **Status**: Pending
+* **Author**: lakshmimsft
+
+## Overview
+
+We've described the design for support of multiple Terraform Providers in Radius in the following PR: [Design Document to support multiple Terraform Providers](https://github.com/radius-project/design-notes/pull/39/files) and implemented secret handling for Private Terraform modules [Ref PR link](https://github.com/radius-project/radius/pull/7306).
+This document  describes in detail the handling of secrets data between the Engine and Driver for Terraform Providers building on the above designs.
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:9]: This overview is hard to parse without jumping through multiple links. Can you instead summarize the purpose of this doc in words and then add links at the end for anyone interested?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 30]
```diff
@@ -0,0 +1,263 @@
...
+| -------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| Terraform Provider | A Terraform provider is a plugin that allows Terraform to interact with and manage resources of a specific infrastructure platform or service, such as AWS, Azure, or Google Cloud. |
+
+## Objectives
+
+**Reference for new type recipeConfig and handling of secrets for Private Terraform Repository:** [Design document for Private Terraform Repository](https://github.com/radius-project/design-notes/blob/3644b754152edc97e641f537b20cf3d87a386c43/recipe/2024-01-support-private-terraform-repository.md)
+
+**Secret Implementation for Private Terraform Repository:**[Secret Implementation for Private Terraform Repository](https://github.com/radius-project/radius/pull/7306)
+
+**Reference for support of multiple Terraform Providers in Radius:** [Design Document to support multiple Terraform Providers](https://github.com/radius-project/design-notes/pull/39/files)
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:23]: Why do we have references in objectives section?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 31]
```diff
@@ -0,0 +1,263 @@
...
+
+> **Issue Reference:** <!-- (If appropriate) Reference an existing issue that describes the feature or bug. -->
+https://github.com/radius-project/radius/issues/6539
+
+### Goals
+
+Describe data flow, specific function calls when using secrets to configure Terraform Providers, building on design in the referred design documents.
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:30]: Can you add user facing goals relevant for the feature rather than goals for what this document describes?

File: recipe/2024-04-terraform-provider-secrets.md [hunk 32]
```diff
@@ -0,0 +1,263 @@
...
+
+### Non goals
+
+The document is focussed on handling secrets for Terraform Providers, with `Applications.Core/secretStores` being the underlying resource for the secret store (which today supports Kubernetes secrets). Other sources and types of `secretsStores` are out of scope for this document.
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:34]: The first half of this sounds more appropriate for goals section rather than non-goal. The second part seems appropriate here but could use more details for clarity.

File: recipe/2024-04-terraform-provider-secrets.md [hunk 33]
```diff
@@ -0,0 +1,263 @@
...
+### Non goals
+
+The document is focussed on handling secrets for Terraform Providers, with `Applications.Core/secretStores` being the underlying resource for the secret store (which today supports Kubernetes secrets). Other sources and types of `secretsStores` are out of scope for this document. 
+
+### User scenarios (optional)
+
+#### User story 1
+As an operator, I maintain a set of Terraform recipes for use with Radius. I have a set of provider configurations which are applicable across multiple recipes. These provider configurations now include secrets, which are injected into the Terraform configuration, allowing the Terraform process to access them as needed.
+
+
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:41]: extra line

File: recipe/2024-04-terraform-provider-secrets.md [hunk 34]
```diff
@@ -0,0 +1,263 @@
...
+## User Experience (if applicable)
+
+**Sample Input:**
+
+This is pulled in from existing document:
+[Design Document to support multiple Terraform Providers](https://github.com/radius-project/design-notes/pull/39/files)
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:47]: Is review this doc required for someone to understand the changes in this section? If not, then I'd remove this reference and call out the additions being proposed in this doc to make it easy to understand.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 20 above)
Comment (kachawla) [reply recipe/2024-04-terraform-provider-secrets.md:234]: Will wait for this to be updated. It's unclear how sensitive flag in the recipe impacts what Radius does here.

File: recipe/2024-04-terraform-provider-secrets.md [hunk 35]
```diff
@@ -0,0 +1,263 @@
...
+
+## Development plan
+Updates to Engine, Driver function call - (under 0.5 Sprint)
+Unit Tests + Functional Test - (under 0.5 Sprint)
+
+Total: 1 Sprint
+
+## Open Questions
+
+We are expecting information on detailed scenarios from PMs on client use cases for Recipe Specific Configuration versus Env Level Provider Config which may get us to revisit design around passing secret data to the driver.
```
Comment (kachawla) [inline recipe/2024-04-terraform-provider-secrets.md:252]: When is this expected? Is this being tracked somewhere?

File: recipe/2024-04-terraform-provider-secrets.md (hunk 20 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md]: I tested this and removed this sentence. yes, Radius can work with variables marked sensitive but there is no impact to what Radius does.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 29 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md]: updated overview. pls take another look.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 30 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md]: Removed these. I think there was an issue number reference that I added to which is not needed.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 31 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md]: updated goals. pls take another look

File: recipe/2024-04-terraform-provider-secrets.md (hunk 32 above)
Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md]: updated. let me know if there is more details i may not be aware of.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 12 above)
Comment (willtsai) [reply recipe/2024-04-terraform-provider-secrets.md:64]: I'm not convinced that we need a CLI experience for this, so I vote we wait for user feedback to dictate whether we should build it.

File: recipe/2024-04-terraform-provider-secrets.md (hunk 35 above)
Comment (willtsai) [reply recipe/2024-04-terraform-provider-secrets.md:244]: is there a reason why we need recipe specific configs? i.e. i haven't heard of users asking for this, has anyone else heard of demand for this?

Comment (lakshmimsft) [reply recipe/2024-04-terraform-provider-secrets.md:244]: With the current design we configure TF providers at the env level. During a recipe execution, data will be processed for all these provider configurations including secrets and passed to the Driver.
During design discussions, it was brought up that it's possible for a specific recipe execution to not need all providers configured and therefore all data sent across and it led to possibility of a more granular approach - recipe specific provider configs.
We were looking to get some more input from PMs after discussions with clients on their user scenarios. As you mention, this is currently not being requested, I can move this out of Open Questions to Design Review Notes and we can revisit this in the future based on feedback.

**Review by rynowak — APPROVED** [review]
